hpx/runtime/components/client_base.hpp

```diff
@@ -155,6 +155,64 @@ namespace hpx { namespace lcos { namespace detail
     };
 }}}
 
+namespace hpx { namespace lcos { namespace detail
+{
+    // Specialization for shared state of id_type, additionally (optionally)
+    // holds a registered name for the object it refers to.
+    template <>
+    struct future_data<id_type> : future_data_base<id_type>
+    {
+        HPX_NON_COPYABLE(future_data);
+
+        using init_no_addref =
+            typename future_data_base<id_type>::init_no_addref;
+
+        future_data() = default;
+
+        future_data(init_no_addref no_addref)
+          : future_data_base<id_type>(no_addref)
+        {}
+
+        template <typename Target>
+        future_data(Target && data, init_no_addref no_addref)
+          : future_data_base<id_type>(std::move(data), no_addref)
+        {}
+
+        future_data(std::exception_ptr const& e, init_no_addref no_addref)
+          : future_data_base<id_type>(e, no_addref)
+        {}
+        future_data(std::exception_ptr && e, init_no_addref no_addref)
+          : future_data_base<id_type>(std::move(e), no_addref)
+        {}
+
+        ~future_data() noexcept override
+        {
+            if (!registered_name_.empty())
+            {
+                std::string name = std::move(registered_name_);
+                error_code ec(lightweight);
+                agas::unregister_name(launch::sync, name, ec);
+            }
+        }
+
+        std::string const& get_registered_name() const override
+        {
+            return registered_name_;
+        }
+        void register_as(std::string const& name, id_type const& id,
+            bool manage_lifetime) override
+        {
+            HPX_ASSERT(registered_name_.empty());   // call only once
+            registered_name_ = name;
+            if (manage_lifetime)
+                hpx::agas::register_name(launch::sync, name, id);
```

Comment:
The shared state should hold the id we want to register the name, right? Why do we need to pass in another one?

Reply:
Any suggestions?

Reply:
Done.